File: JavaAlgorithm/src/main/java/algo/tzashinorpu/ThirdRound/Chapter09/numIslands_200.java
package algo.tzashinorpu.ThirdRound.Chapter09;

public class numIslands_200 {
    public int numIslands(char[][] grid) {
        int[][] dirs = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        int cnt = 0;
        for (int i = 0; i < grid.length; i++) {
            for (int j = 0; j < grid[i].length; j++) {
                if (grid[i][j] == '1') {
                    cnt++;
                    _dfs(i, j, grid, dirs);
                }
            }
        }
        return cnt;
    }

    private void _dfs(int i, int j, char[][] grid, int[][] dirs) {
        if (invalid(i, j, grid)) {
            return;
        }
        if (grid[i][j] != '1') {
            return;
        }
        grid[i][j] = 'x';
        for (int k = 0; k < 4; k++) {
            _dfs(i + dirs[k][0], j + dirs[k][1], grid, dirs);
        }
    }

    private boolean invalid(int i, int j, char[][] grid) {
        return i < 0 || i >= grid.length || j < 0 || j >= grid[0].length;
    }
}

File: Clion/ZeroTrac/Graph/numIslands_200.cpp
#include "../header/g"
using namespace G;

int Solution::numIslands(vector<vector<char>> &grid)
{

    int res = 0;
    int m = grid.size(), n = grid[0].size();
    vector<vector<int>> dirs = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
    function<void(vector<vector<char>> &, int, int)> dfs = [&](vector<vector<char>> &grid, int x, int y) -> void
    {
        if (x < 0 || x == m || y < 0 || y == n || grid[x][y] == '0')
            return;
        grid[x][y] = '0';
        for (auto dir : dirs)
            dfs(grid, x + dir[0], y + dir[1]);
    };
    for (int i = 0; i < m; i++)
        for (int j = 0; j < n; j++)
            if (grid[i][j] == '1')
                res++, dfs(grid, i, j);
    return res;
}